### PR TITLE
New methods: concat, fill, join

### PR DESCRIPTION
**Pull request by BabkinAleksandr**
Concat - 308 B
Concats array-... to array, string-... to string

Fill - 296 B

Join - 278 B

### CODE REVIEW

File: lib/concat/index.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import curry from '../curry'
+
+export default curry(function concat(arg) {
+  return [].slice.call(arguments, 1).reduce(function(acc, item) {
```
Comment (Beraliv) [inline lib/concat/index.js]: toArray here

File: lib/concat/concat.test.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+var concat = require('.')
+
+const arr = [1, 2, 3, 4, 5];
+const str = 'hello';
+
+test("concats array with any number of arguments", () => {
+  expect(concat(arr, 'a')).toEqual([...arr, 'a'])
+  expect(concat(arr, 'a', 'b')).toEqual([...arr, 'a', 'b'])
+  expect(concat(arr, 'a', 'b', 'c')).toEqual([...arr, 'a', 'b', 'c'])
+  expect(concat(arr, 'a', 'b', 'c', 'd')).toEqual([...arr, 'a', 'b', 'c', 'd'])
```
Comment (Beraliv) [inline lib/concat/concat.test.js]: doesn't concat alway return a string?

File: lib/fill/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import curry from '../curry'
+
+export default curry(function fill(arr, value, start, end) {
```
Comment (Beraliv) [inline lib/fill/index.js]: will better use curryN(4, ...)
check the size, it will be less than current

File: lib/fill/index.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  return Array.isArray(arr) ?
+    value ?
```
Comment (Beraliv) [inline lib/fill/index.js]: put ? to a new line

File: lib/concat/concat.test.js (hunk 2 above)
Comment (BabkinAleksandr) [reply lib/concat/concat.test.js]: According to [ramda concat](http://ramdajs.com/docs/#concat) - not always

File: lib/fill/index.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+      arr.fill(value, start, end)
+      :
```
Comment (Beraliv) [inline lib/fill/index.js]: align with appropriate `?`

File: lib/fill/index.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+      arr
+    :
```
Comment (Beraliv) [inline lib/fill/index.js]: here as well

File: lib/join/index.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import curry from '../curry'
+
+export default curry(function join(d, arr) {
```
Comment (Beraliv) [inline lib/join/index.js]: curryN(2, ...) will take less memory
and check the size adding join and curry at the same file if it's worth it (i still don't know)

File: lib/concat/concat.test.js (hunk 2 above)
Comment (Beraliv) [reply lib/concat/concat.test.js]: If both arguments are strings, the result will be a string, right

**Comment by BabkinAleksandr** [comment]
Concat - 295 B
Fill - 284 B
Join - 267 B (curried), 33 B (not curried)

Also, forgot to say: curried fill method fails all tests, when not all args is given, for example:
```js
fill(arr, '0') // start, end not provided
```
Error: 
babkin_mbp-admin____documents_nanoutils__zsh_
function return function

Do not understand, is it my mistake?

**Comment by Kelin2025** [comment]
Because `fill` has 4 arguments and curried with 4, `curried` function waits until all 4 arguments will be passed

File: lib/fill/index.js [hunk 8]
```diff
@@ -0,0 +1,9 @@
+import curryN from '../curryN'
+
+export default curryN(4, function fill(arr, value, start, end) {
```
Comment (Beraliv) [inline lib/fill/index.js]: you can set start = 0 and end = arr.length -1 and then use curryN(2, ...) but you won't be able to specify start and end anymore

As a solution, you can use not a value, but an object 
```
{
  value,
  start = 0,
  end = arr.length - 1
}
```

Comment (Kelin2025) [reply lib/fill/index.js]: It's a part of ES6, we use ES5 only while we have Node LTS 4/6

Comment (Beraliv) [reply lib/fill/index.js]: yeah, I mean object as a second parameter and then parse it

But better idea will be to remove start and end at all
It's a different method (like slice, you can use it after in chain)

Comment (BabkinAleksandr) [reply lib/fill/index.js]: So, should I just remove them?

**Comment by Kelin2025** [comment]
Also I haven't found `fill` method in Ramda, is it there or not?

**Comment by BabkinAleksandr** [comment]
I used[ lodash fill](https://lodash.com/docs/4.17.4#fill) as reference

**Comment by Kelin2025** [comment]
Ok, but if you'll find Ramda's equivalent with another name, use name from Ramda
We want maximum of compability with Ramda

File: lib/fill/index.js (hunk 8 above)
Comment (Beraliv) [reply lib/fill/index.js]: Yeah, and place 2 instead of 4 then

**Comment by BabkinAleksandr** [comment]
Ok, I used lodash because of _Post Scriptum_ of [your article](https://medium.com/@kelin2025/writing-js-libraries-less-than-1tb-size-6342da0c006a), where you mention it:

writing_js_libraries_less_than_1tb_size_ _anton_kosykh_ _medium

Now I will use Ramda as reference.

**Comment by BabkinAleksandr** [comment]
Fill - 281 B

File: lib/concat/index.js.flow [hunk 9]
```diff
@@ -0,0 +1,2 @@
+// @flow
+declare module.exports: (arg: Array<mixed>|string, ...values: Array<mixed>|string) => Array<mixed>|string
```
Comment (Kelin2025) [inline lib/concat/index.js.flow]: Incorrect type definition for curried function

File: lib/fill/index.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import curryN from '../curryN'
+
+export default curryN(2, function fill(arr, value) {
+  return Array.isArray(arr)
+    ? value
+      ? arr.fill(value)
```
Comment (Kelin2025) [inline lib/fill/index.js]: It mutates arr

File: lib/fill/index.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import curryN from '../curryN'
+
+export default curryN(2, function fill(arr, value) {
+  return Array.isArray(arr)
```
Comment (Kelin2025) [inline lib/fill/index.js]: I think no need to check if arr is Array - just let it fall

File: lib/fill/index.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import curryN from '../curryN'
+
+export default curryN(2, function fill(arr, value) {
```
Comment (Kelin2025) [inline lib/fill/index.js]: [`_.fill`](https://lodash.com/docs/4.17.4#fill) also has `start` and `end` args to fill only specified range  
Leave `curryN(2)`, if you'll add `start` and `end`

File: lib/concat/index.js.flow (hunk 9 above)
Comment (Kelin2025) [inline lib/concat/index.js.flow]: Incorrect type definition for curried function

File: lib/fill/index.js.flow [hunk 13]
```diff
@@ -0,0 +1,2 @@
+// @flow
+declare module.exports: (arr: Array<?mixed>, value: ?string, start: ?number, end: ?number) => Array<?mixed>
```
Comment (Kelin2025) [inline lib/fill/index.js.flow]: Incorrect type definition for curried function

File: lib/fill/index.js (hunk 12 above)
Comment (BabkinAleksandr) [reply lib/fill/index.js]: Sorry, do not understand what do you mean. We need to return back _start_ and _end_ args, but leave curryN(2)?

File: lib/fill/index.js (hunk 10 above)
Comment (BabkinAleksandr) [reply lib/fill/index.js]: Original lodash method mutates array, but I think not mutating is better

lodash_documentation

File: lib/fill/index.js (hunk 12 above)
Comment (Kelin2025) [reply lib/fill/index.js]: `curryN(2, function fill(arr, value, start, end) {})`
`curryN(2)` allows use `fill(arr)(value)` without start/end + use `fill(arr)(value, start)` to use without end + use `fill(arr)(value, start, end)` to use without end  
But I think the better idea will be:
`fill(val)(arr)`
`fill(val, start)(arr)`
`fill(val, start, end)(arr)`
Still needs `curryN(2)` but also needs to check where is arr (just with `arguments.length`)

File: lib/fill/index.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import curryN from '../curryN'
+import toArray from '../toArray'
+
+export default curryN(2, function fill(val, start, end, arr) {
+  var last = arguments[arguments.length - 1]
+  return Array.isArray(last)
+    ? val
+      ? toArray(last).fill(val, Array.isArray(start) ? undefined : start, Array.isArray(end) ? undefined : end)
```
Comment (Beraliv) [inline lib/fill/index.js]: repepetive code, extract it to a function

File: lib/fill/index.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+      : last
+    : function(arrC) {
+        return fill(val, start, end, arrC)
+      }
```
Comment (Beraliv) [inline lib/fill/index.js]: what does it mean? function returns function? it's not HOC.

Comment (BabkinAleksandr) [reply lib/fill/index.js]: as @Kelin2025 said in https://github.com/Kelin2025/nanoutils/pull/14#discussion_r164082942 we expect function could be called like:
`fill(val)(arr)`
`fill(val, start)(arr)`
`fill(val, start, end)(arr)`
`fill(val, start, end, arr)`

So, if `arr` was provided through currying, we should return function to take `arr` as argument and put it in new `fill`

Maybe we do not need `curryN(2,...)` wrapper in this situation? It's size would be 142B

File: lib/fill/index.js (hunk 14 above)
Comment (BabkinAleksandr) [reply lib/fill/index.js]: agree

File: lib/concat/index.js [hunk 16]
```diff
@@ -0,0 +1,6 @@
+import curryN from '../curryN'
+import toArray from '../toArray'
+
+export default curryN(2, function concat(arg) {
+  return arg.concat.apply(arg, toArray(arguments).slice(1))
```
Comment (Beraliv) [inline lib/concat/index.js]: here slice is repeated twice

File: lib/fill/index.js (hunk 14 above)
Comment (Beraliv) [reply lib/fill/index.js]: Validate it as a number only, why do you need to check is it's an array?

File: lib/fill/index.js (hunk 15 above)
Comment (Beraliv) [reply lib/fill/index.js]: Yeah, it might be reasonable

Comment (Beraliv) [reply lib/fill/index.js]: can we use bind here? Will it be better?

Comment (Kelin2025) [reply lib/fill/index.js]: I know easy solution, I'll publish it later

File: lib/fill/fill.test.js [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+test("return array itself if value is undefined", () => {
+  const arr = [1, 2, 3, 4, 5]
+  expect(fill(undefined, arr)).toEqual(arr)
+})
+
+test("fills array with value from start index", () => {
+  const arr = [1, 2, 3, 4, 5];
+  expect(fill('*', 2)(arr)).toEqual([1, 2, '*', '*', '*'])
+  expect(fill('a', 4, arr)).toEqual([1, 2, 3, 4, 'a'])
```
Comment (Beraliv) [inline lib/fill/fill.test.js]: oh, no! you're mutating an array here.

File: lib/fill/fill.test.js [hunk 18]
```diff
@@ -0,0 +1,24 @@
+var fill = require('.')
+
+test("fills array with value", () => {
+  const arr = [1, 2, 3, 4, 5]
+  expect(fill('*', arr)).toEqual(['*', '*', '*', '*', '*'])
+  expect(fill('a')(arr)).toEqual(['a', 'a', 'a', 'a', 'a'])
```
Comment (Beraliv) [inline lib/fill/fill.test.js]: mutating an array here

File: lib/fill/index.js (hunk 14 above)
Comment (Beraliv) [inline lib/fill/index.js]: no, no and again no
image

**Pull request merged**